ding/policy/ppg.py

```diff
@@ -34,6 +34,589 @@ def create_shuffled_dataloader(data, batch_size):
+                     objective while otherwise preserving the original policy
+        """
+        data = self._data_preprocess_learn(data)
+        # ====================
+        # PPG forward
+        # ====================
+        self._learn_model.train()
+        return_infos = []
+        if self._value_norm:
+            unnormalized_return = data['adv'] + data['value'] * self._running_mean_std.std
+            data['return'] = unnormalized_return / self._running_mean_std.std
+            self._running_mean_std.update(unnormalized_return.cpu().numpy())
+        else:
+            data['return'] = data['adv'] + data['value']
+
+        for epoch in range(self._cfg.learn.actor_epoch_per_collect):
+            for batch in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):
+                policy_data = batch
+                policy_adv = policy_data['adv']
+                if self._adv_norm:
+                    # Normalize advantage in a total train_batch
+                    policy_adv = (policy_adv - policy_adv.mean()) / (policy_adv.std() + 1e-8)
+                # Policy Phase(Policy)
+                policy_output = self._learn_model.forward(policy_data['obs'], mode='compute_actor')
+                policy_error_data = ppo_policy_data(
+                    policy_output['logit'], policy_data['logit'], policy_data['action'], policy_adv,
+                    policy_data['weight']
+                )
+                ppo_policy_loss, ppo_info = ppo_policy_error(policy_error_data, self._clip_ratio)
+                policy_loss = ppo_policy_loss.policy_loss - self._entropy_weight * ppo_policy_loss.entropy_loss
+                self._optimizer_ac.zero_grad()
+                policy_loss.backward()
+                self._optimizer_ac.step()
+
+        for epoch in range(self._cfg.learn.critic_epoch_per_collect):
+            for batch in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):
+                value_data = batch
+                value_adv = value_data['adv']
+                return_ = value_data['return']
+                if self._adv_norm:
+                    # Normalize advantage in a total train_batch
+                    value_adv = (value_adv - value_adv.mean()) / (value_adv.std() + 1e-8)
+                # Policy Phase(Value)
+                value_output = self._learn_model.forward(value_data['obs'], mode='compute_critic')
+                value_error_data = ppo_value_data(
+                    value_output['value'], value_data['value'], return_, value_data['weight']
+                )
+                value_loss = self._value_weight * ppo_value_error(value_error_data, self._clip_ratio)
+                self._optimizer_aux_critic.zero_grad()
+                value_loss.backward()
+                self._optimizer_aux_critic.step()
+
+        data['return_'] = data['return']
+
+        self._aux_memories.append(copy.deepcopy(data))
```

Comment:
test whether we can remove `deepcopy` here

Reply:
No, learn_aux was not always 1